packages/core/database/src/index.ts

```diff
@@ -139,7 +139,7 @@ class Database {
   }
 
   getSchemaName(): string | undefined {
-    return this.connection.client.connectionSettings.schema;
+    return this.connection.client?.connectionSettings?.schema;
```

Comment:
What about?

```js
return this.connection.client?.connectionSettings?.schema || "public";
```

Or fill it in with an optional environment variable?

That would at least make this functionality available in the near future. The unset default is already "public".

Reply:
cc @innerdvations

Reply:
Sorry, it's been a while so I might be misunderstanding, but if connection is a function that uses a schema, won't this then mistarget "public"?

I totally agree this is a flaw in our implementation, we should not be accessing connection.client.connectionSettings directly at all and plan to remove this method entirely in Strapi v5. But currently we have a freeze on v4, and changing this to support a connection function could introduce some subtle issues we don't want to have to solve in v4.

Reply:
> Sorry, it's been a while so I might be misunderstanding, but if connection is a function that uses a schema, won't this then mistarget "public"?
> 
> I totally agree this is a flaw in our implementation, we should not be accessing connection.client.connectionSettings directly at all and plan to remove this method entirely in Strapi v5. But currently we have a freeze on v4, and changing this to support a connection function could introduce some subtle issues we don't want to have to solve in v4.

It would. That's why I suggested you could use an env variable. For example:

```js
return this.connection.client?.connectionSettings?.schema || process.env?.DATABASE_SCHEMA || "public";
```

From what I can tell the connection function would simply be returning the identical schema anyway.

Here is the use case:

```js
// By default, the configuration object received via a function is cached and reused for all connections. To change this 
// behavior, an expirationChecker function can be returned as part of the configuration object. The expirationChecker 
// is consulted before trying to create new connections, and in case it returns true, a new configuration object is
// retrieved. For example, to work with an authentication token that has a limited lifespan:

const knex = require('knex')({
  client: 'postgres',
  connection: async () => {
    const { 
      token, 
      tokenExpiration 
    } = await someCallToGetTheToken();

    return {
      host : 'your_host',
      port : 5432,
      user : 'your_database_user',
      password : token,
      database : 'myapp_test',
      expirationChecker: () => {
        return tokenExpiration <= Date.now();
      }
    };
  }
});
```

The way to use it with Strapi so that I can access the Managed Identity token inside the App Service (so that I can have a passwordless connection and don't have to manage the secrets) and also so that it can be renewed when it expires:

```js
import { DefaultAzureCredential } from "@azure/identity";

export default ({ env }) => {
  const connections = {
    postgres: {
      connection: async () => {
        const credential = new DefaultAzureCredential();
        const accessToken = await credential.getToken('https://ossrdbms-aad.database.windows.net/.default');

        return {
          connectionString: env('DATABASE_URL'),
          host: env('DATABASE_HOST', 'servername.postgres.database.azure.com'),
          port: env.int('DATABASE_PORT', 5432),
          database: env('DATABASE_NAME', 'strapi'),
          user: env('DATABASE_USERNAME', account@domain.onmicrosoft.com'),
          password: accessToken.token,
          expirationChecker: () => {
            return accessToken.expiresOnTimestamp <= Date.now();
          },
          ssl: env.bool('DATABASE_SSL', false) && {
            key: env('DATABASE_SSL_KEY', undefined),
            cert: env('DATABASE_SSL_CERT', undefined),
            ca: env('DATABASE_SSL_CA', undefined),
            capath: env('DATABASE_SSL_CAPATH', undefined),
            cipher: env('DATABASE_SSL_CIPHER', undefined),
            rejectUnauthorized: env.bool(
              'DATABASE_SSL_REJECT_UNAUTHORIZED',
              true
            ),
          },
          schema: env('DATABASE_SCHEMA', 'public')
        }
      },
      pool: { min: env.int('DATABASE_POOL_MIN', 2), max: env.int('DATABASE_POOL_MAX', 10) },
    }
 }
```

From eyeballing, I'm not seeing anything in the code that breaks and I've stood up an instance that I've tested lightly with this change. Would your e2e tests not catch anything strange with this?

Sidenote: when would 5 be coming out and would 5 properly expose the knex API?